lnwallet/chanfunding/wallet_assembler.go

```diff
@@ -488,7 +488,13 @@ func (w *WalletAssembler) ProvisionChannel(r *Request) (Intent, error) {
 		for _, coin := range selectedCoins {
 			outpoint := coin.OutPoint
 
-			w.cfg.CoinLocker.LockOutpoint(outpoint)
+			_, _, _, err = w.cfg.CoinLeaser.LeaseOutput(
+				LndInternalLockID, outpoint,
+				DefaultLockDuration,
```

Comment:
I wonder: Since before this change the lock would basically exist forever (or until the next restart of the node), should we also use an "infinite" duration here? Or do we kind of rely on the general funding timeout we have in `lnd` that is 10 minutes?

Reply:
I think we have an interplay of several factors:

- the `chanfunding.DefaultLockDuration` (formerly `walletrpc.DefaultLockDuration`) is 10m
- the `lncfg.DefaultReservationTimeout` is 10m
- only the dev config allows changing the reservation timeout from 10m, which is used in itests
- the reservation timeout applies to time since last update to the reservation, rather than time since the reservation was created - this means it can go longer than 10m if the peer is slow enough to respond
- the zombie sweeper iterates through `activeReservations` which I believe is NOT populated on startup from any non-volatile data, so it would miss releasing outputs with infinite expiration and would cause them to need to be released manually in the case that LND restarts with an active reservation

This makes me think that maybe the duration should be the reservation timeout multiplied by a constant factor for the number of state transitions we expect, to allow the peer to idle up to the timeout at each step. Since the timeout is configurable in the dev environment, perhaps we pass that value to lnwallet when initializing it as well?

Reply:
Ah, right, we have the zombie sweeper that does the cleanup. Maybe we could add a `Cleanup()` method to the reservation that would be called by the zombie sweeper to release the locks immediately? That combined with a multiplication factor of the reservation timeout sounds good to me.

Reply:
- The zombie sweeper calls `failFundingFlow()` on the manager with the channel ID here: https://github.com/lightningnetwork/lnd/blob/72764b14734870d03a1212795e2ec7c02f3b4500/funding/manager.go#L4848
- which calls `cancelReservationCtx()` on the manager with the channel ID here: https://github.com/lightningnetwork/lnd/blob/72764b14734870d03a1212795e2ec7c02f3b4500/funding/manager.go#L912
- which calls `Cancel()` on the `ChannelReservation` here: https://github.com/lightningnetwork/lnd/blob/72764b14734870d03a1212795e2ec7c02f3b4500/funding/manager.go#L4885
- which sends a `fundingReserveCancelMsg` to the wallet's `msgChan` with the reservation ID here: https://github.com/lightningnetwork/lnd/blob/72764b14734870d03a1212795e2ec7c02f3b4500/lnwallet/reservation.go#L795
- which calls `handleFundingCancelRequest()` on the wallet with the message here: https://github.com/lightningnetwork/lnd/blob/72764b14734870d03a1212795e2ec7c02f3b4500/lnwallet/wallet.go#L620
- which calls `UnlockOutpoint()` here: https://github.com/lightningnetwork/lnd/blob/72764b14734870d03a1212795e2ec7c02f3b4500/lnwallet/wallet.go#L1428
- which is replaced in this PR already with a `ReleaseOutput()`

So I think the `Cleanup()` part of this might be redundant, as the zombie sweeper already calls cleanup code that this PR modifies to work with leases instead of locks. I think all I need to do is update this timeout.

Reply:
Okay cool. Sounds good to me, thanks for confirming!

Reply:
Thanks! I got most of the way there today, will finish it up tomorrow.

Reply:
Looking more closely at the multiplier to use:

- Channel establishment v1 should have a multiplier of 1 (the time between `funding_created` and `funding_signed` messages)
- Channel establishment v2 specifically recommends not locking UTXOs and reusing them between channel open attempts for dual-funded channels, to help [prevent liquidity griefing](https://github.com/lightning/bolts/blob/master/02-peer-protocol.md#liquidity-griefing). In addition, since there's a loop that's iterated for each input added (from each side), the multiplier isn't constant.
 
For itests, the reservation timeout is shorter, so using the longer value of the default timeout is safe. I think it's safe to just use `lncfg.DefaultReservationTimeout` in this case.

Reply:
Moved `lncfg.DefaultReservationTimeout` to `lnwallet/chanfunding` to avoid import cycles, updated all references to it, and used it explicitly instead of `DefaultLockDuration` for channel reservations.

Reply:
Ah, forgot to actually use `DefaultReservationTimeout` instead of `DefaultLockDuration`. Fixed.